### PR TITLE
[stdlib] Optimize some Set operations

### PR DESCRIPTION
**Pull request by lorentey**
This uses temporary bitmaps to considerably speed up some high-level Set operations.

Operation | Speedup
---|---
`Set.isSubset<S>(of:)` | 3x
`Set.isStrictSubset<S>(of:)` | 3x
`Set.isSuperset<S>(of:)` | 4x
`Set.isStrictSuperset<S>(of:)` | 4x
`Set.isDisjoint<S>(with:)` | 4x
`Set.subtracting(_:)` | 1-4x
`Set.filter(_:)` | 1-4x
`Set.intersection<S>(_:)` | 4-6x
`Set.intersection(_:)` | 1-4x

Benchmark results depend on the size of the sets and the element type. The numbers above were measured with integer keys across a range of counts from 1 to 1 million items.

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
The benchmark suit doesn't cover all these operations. Here are some highlights from local testing:

`Set.isSubset<S: Sequence>(of: S)`:
set - 8 tasks

`Set.isStrictSubset<S: Sequence>(of: S)`:
set - 8 tasks

`Set.isSuperset<S: Sequence>(of: S)`:
set - 8 tasks

`Set.isStrictSuperset<S: Sequence>(of: S)`:
set - 8 tasks

`Set.isDisjoint<S: Sequence>(with: S)`:
set - 8 tasks

**Comment by lorentey** [comment]
Similar charts for operations returning sets:

`Set.subtracting(_: Set)`:
set - 10 tasks

`Set.subtracting<S: Sequence>(_: S)`:
set - 10 tasks

`Set.filter(_:)`:
set - 10 tasks

`Set.intersection(_: Set)`:
set - 10 tasks

`Set.intersection<S: Sequence>(_: S)`:
set - 10 tasks

File: stdlib/public/core/Bitset.swift [hunk 1]
```diff
@@ -337,3 +337,250 @@ extension _UnsafeBitset.Word: Sequence, IteratorProtocol {
...
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+extension _Bitset {
+  /// A simple bitmap storage class with room for a specific number of
+  /// tail-allocated bits.
+  @_fixed_layout
+  @usableFromInline
+  internal final class Storage {
```
Comment (jrose-apple) [inline stdlib/public/core/Bitset.swift]: Why isn't this a ManagedBuffer?

Comment (lorentey) [reply stdlib/public/core/Bitset.swift]: Hm, maybe it should be! I tend to avoid it when I can help it: I find its API unappealing.

Comment (jrose-apple) [reply stdlib/public/core/Bitset.swift]: That's a comment that probably deserves some consideration. :-) But any bare manipulation of tail elements scares me, so…

Comment (lorentey) [reply stdlib/public/core/Bitset.swift]: The really scary part is all the unsafe access, and ManagedBuffer doesn't help with those. On the other hand, it adds extra layering and pollutes the class interface, which isn't always helpful.

I did get burnt by rolling my own header initialization before, which isn't an issue with ManagedBuffer. I'll have a look, and if it's not too bad, I'll switch to it!

**Comment by lorentey** [comment]
@swift-ci please test linux platform

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
@swift-ci please test

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by lorentey** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please test

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
The runtimes of the extended `Set.Sequence` benchmarks from #24156 look good (all in the 20–1000 μs range). PR #24725 disables 3 benchmarks that run too long with the current implementation. Please revert it to re-enable them shortly before these optimizations land.

Similarly, there is a small issue with too low runtime (even before these changes) in a case of `Set.subtracting.Empty.Box`. This would be fixed by #24570, which should IMHO be also merged shortly before this one, to batch all the changes to benchmark runtimes without needing to burn good names on a small bug fix.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Comment by lorentey** [comment]
Closing in favor of #40012.

**Pull request closed without merging**